### DISCUSSION TITLE
A way to get blob to use UInt16List rather than UInt8List

### DISCUSSION THREAD

**Discussion by JerContact** [simolus3/drift#2523] [Q&A]
I'm trying to add a List<int> as a blob into the sqlite database, but noticed when I use blob it is only using UInt8List.  I would need to manually convert all the numbers in the List<int> to match UInt8List so that it can go higher than the limit of int8, so was just wondering if there was a way to have the blob be UInt16List as an option?  Or, if you have any suggestions!

```
CREATE TABLE Test (
    id INTEGER PRIMARY KEY,
    Children BLOB,
);
```

**Comment by simolus3** [comment]
It's not a builtin option, since `BLOB` kind of describes bytes, right? There's also an endianness question if you create a database on a little-endian device and later copy it to a big-endian computer. If you don't care about that, you could write a [type converter](https://drift.simonbinder.eu/docs/advanced-features/type_converters/) like this:

```dart
class BlobAsUint16ListConverter extends TypeConverter<Uint16List, Uint8List> {
  const BlobAsUint16ListConverter();

  @override
  Uint16List fromSql(Uint8List fromDb) {
    return fromDb.buffer.asUint16List(fromDb.offsetInBytes, fromDb.length ~/ 2);
  }

  @override
  Uint8List toSql(Uint16List value) {
    return value.buffer.asUint8List(value.offsetInBytes, value.length * 2);
  }
}
```

You can then apply it like this:

```sql
import 'that_converter.dart';

CREATE TABLE Test (
    id INTEGER PRIMARY KEY,
    Children BLOB MAPPED BY `const BlobAsUint16ListConverter()`,
);
```

Or, if you're using a Dart table definition, `=> blob().map(const BlobAsUint16ListConverter())()`.

**Reply by JerContact** [reply]
Yeah, I guess from a small bit of reading I thought trying to encode it and decode it using the same byte buffer that it would take care of the endian question, but maybe not?  I might need to read more about that.

This option is great though!!!  Thanks!

**Reply by simolus3** [reply]
> thought trying to encode it and decode it using the same byte buffer that it would take care of the endian question, but maybe not?

This is not actually encoding or decoding something. A `Uint8List`, just like a `Uint16List` and all the other typed lists, are just a view over a buffer in memory. So when you have a uint16 list on a little-endian machine over a memory buffer starting at byte address `a`, doing `thatList[5] = 0x1234` will write `0x34` to address `a + 10` and `0x12` to address `a + 11`. But the database views this list as a range of bytes. So if you later read from the database with a big-endian machine, `thatList[5]` will be expect `a + 10` to hold the high bytes and `a + 11` to hold the lower bytes. So you'd get `0x3412` back.

It's pretty unlikely your database will move to a big-endian device, but it's something to keep in mind.
